apps/prairielearn/src/ee/lib/validateHTML.ts

```diff
@@ -0,0 +1,616 @@
+import * as parse5 from 'parse5';
+
+const HTML_GLOBAL_TAGS = [
+  'accesskey',
+  'class',
+  'contenteditable',
+  'data-*',
+  'dir',
+  'draggable',
+  'enterkeyhint',
+  'hidden',
+  'id',
+  'inert',
+  'inputmode',
+  'lang',
+  'popover',
+  'spellcheck',
+  'style',
+  'tabindex',
+  'title',
+  'translate',
+];
+
+/**
+ * Checks that the required property is an int (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param errors The list of errors to add to.
+ */
+function assertInt(tag: string, key: string, val: string, errors: string[]) {
+  if (!(/^\d+$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an integer, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is an float (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param errors The list of errors to add to.
+ */
+function assertFloat(tag: string, key: string, val: string, errors: string[]) {
+  if (!(/^(\d+)\.?(\d*)(e-\d+)?$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an floating-point number, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is in a list of possibilities (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertInChoices(
+  tag: string,
+  key: string,
+  val: string,
+  choices: string[],
+  errors: string[],
+) {
+  if (!(choices.includes(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be in ${choices}, but value provided is ${val}`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is a boolean (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertBool(tag: string, key: string, val: string, errors: string[]) {
+  assertInChoices(tag, key, val, ['true', 'false'], errors);
+}
+
+/**
+ * Checks that a tag has valid properties.
+ * @param ast The tree to consider, rooted at the tag.
+ * @returns The list of errors for the tag, if any.
+ */
+function checkTag(ast: any): string[] {
```

Comment:
Hopefully there's a better type than `any`?

Reply:
There is: `DocumentFragment | ChildNode` from `'parse5/dist/tree-adapters/default'`. For some reason, I can import the types in the dev server but not in the build system.

Reply:
What did you try, what error are you seeing, and when are you seeing that error?

Reply:
Nvm, I found how to get the types out of `parse5.DefaultTreeAdapterMap`. Changed typing in abb2b3997dd8ce8bb963ccb76a0b4b112fdfc733.